src/main/java/uk/gov/justice/digital/zone/StructuredZone.java

```diff
@@ -181,95 +135,57 @@ private void handleInvalidRecords(SparkSession spark,
+            SparkSession spark,
+            Dataset<Row> dataFrame,
+            String destinationPath,
+            SourceReference.PrimaryKey primaryKey
+    ) throws DataStorageException {
+        val validRecords = dataFrame.filter(col(VALID).equalTo(true)).select(PARSED_DATA + ".*", OPERATION);
+        val validRecordsCount = validRecords.count();
 
-        val missingSchemaRecords = dataFrame
-                .select(col(DATA), col(METADATA))
-                .withColumn(ERROR, lit(String.format("Schema does not exist for %s/%s", source, table)));
+        if (validRecordsCount > 0) {
+            logger.info("Writing {} valid records", validRecordsCount);
+            writeValidRecords(spark, storage, destinationPath, primaryKey, validRecords);
 
-        appendDataAndUpdateManifestForTable(
-                spark,
-                missingSchemaRecords,
-                createValidatedPath(violationsPath, source, table)
-        );
-        return createEmptyDataFrame(dataFrame);
+            return validRecords;
+        } else {
+            logger.info("No valid records found");
+            return createEmptyDataFrame(dataFrame);
+        }
     }
 
-    private void appendDataAndUpdateManifestForTable(SparkSession spark,
-                                                     Dataset<Row> dataFrame,
-                                                     String tablePath) throws DataStorageException {
-        logger.info("Appending {} records to deltalake table: {}", dataFrame.count(), tablePath);
-        storage.append(tablePath, dataFrame.drop(OPERATION));
-        logger.info("Append completed successfully");
-        storage.updateDeltaManifestForTable(spark, tablePath);
+    private Dataset<Row> validateJsonData(
+            SparkSession spark,
+            Dataset<Row> dataFrame,
+            StructType schema,
+            String source,
+            String table
+    ) {
+        logger.info("Validating data against schema: {}/{}", source, table);
+        val jsonValidator = JsonValidator.createAndRegister(schema, spark, source, table);
+
+        return dataFrame
+                .select(col(DATA), col(METADATA), col(OPERATION))
+                .withColumn(PARSED_DATA, from_json(col(DATA), schema, jsonOptions))
+                .withColumn(VALID, jsonValidator.apply(col(DATA), to_json(col(PARSED_DATA), jsonOptions)));
     }
 
-    private void appendDataAndUpdateManifestForTable(SparkSession spark,
-                                                     Dataset<Row> dataFrame,
-                                                     String tablePath,
-                                                     SourceReference.PrimaryKey primaryKey) throws DataStorageException {
-        logger.info("Appending {} records to deltalake table: {}", dataFrame.count(), tablePath);
-        storage.appendDistinct(tablePath, dataFrame.drop(OPERATION), primaryKey);
+    @Override
+    public void writeInvalidRecords(
```

Comment:
Ideally this shouldn't be part of the public api. Would probably mean an abstract class instead with protected methods though I think.